python/simulation.py

```diff
@@ -3237,19 +3237,16 @@ def get_source(self, component, vol=None, center=None, size=None):
         self.fields.get_source_slice(v, component ,arr)
         return arr
 
-    # if return_pw, the return value is (points, weights) where points is a
-    # mp.Vector3-valued array of the same dimensions as weights.
-    # otherwise return value is 4-tuple (xtics, ytics, ztics, weights).
     def get_array_metadata(self, vol=None, center=None, size=None, dft_cell=None,
-                           collapse=False, snap=False, return_pw=False):
+                           collapse=False, return_pw=False):
         """
         This routine provides geometric information useful for interpreting the arrays
         returned by `get_array` or `get_dft_array` for the spatial region defined by `vol`
         or `center`/`size`. In both cases, the return value is a tuple `(x,y,z,w)`, where:
 
-        + `x,y,z` are 1d NumPy arrays storing the $x,y,z$ coordinates of the points in the
+        + `x,y,z` are tuples storing the $x,y,z$ coordinates of the points in the
```

Comment:
Just another drive by comment following this from the reference to #1431... `x`, `y`, and `z` look like they're NumPy arrays, not tuples.

Reply:
Thanks for calling this out. The "tics" return values `(x,y,z)` should be NumPy arrays but in fact they are set up as tuples:

https://github.com/NanoComp/meep/blob/f5fc4cd469b9c89759dcfe30a96ca4814eb9458d/python/simulation.py#L3278-L3289

I verified this using `type(x)`, `type(y)`, `type(z)` and the results was a tuple. (`w` though is a NumPy array.) If we really want these return values as 1d NumPy arrays then we will need to modify `get_array_metadata` in a separate PR. I just updated the documentation here to make it consistent with what is actually happening.

Reply:
Ah, good to know. You could also call `np.asarray()` on them to do the conversion in Python.

Reply:
This looks like a SWIG problem — `self.fields.get_array_metadata(v, collapse, snap) ` is nowadays returning a `tuple`, whereas the intention (and I think the original behavior) was to return a NumPy array.

We definitely want a numpy array here — a tuple is a terrible data structure for a large array.  And we want to return a numpy array directly, not return a tuple and then convert to numpy.

Reply:
`fields::get_array_metadata` returns a `std::vector<double>`, which we want SWIG to convert to a NumPy array.  I'm not sure why that isn't happening.

Reply:
According to the documentation for SWIG's [`std_vector.i`](http://www.swig.org/Doc1.3/Library.html#Library_nn15) (which we are using), it's supposed to be using a "list object in the target language" for output typemaps.

Reply:
I was able to fix it by adding:
```cxx
// use NumPy arrays for returning common std::vector types:
%typemap(out) std::vector<double> {
    npy_intp vec_len = (npy_intp) $1.size();
    $result = PyArray_SimpleNew(1, &vec_len, NPY_DOUBLE);
    memcpy(PyArray_DATA((PyArrayObject*) $result), &$1[0], vec_len * sizeof(double));
}
```
to `meep.i` (after `%template(DoubleVector) std::vector<double>;`).

Reply:
We can do that in a separate PR.

Reply:
See #1458.